### ISSUE TITLE
[Phase 7] Integration & Polish

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#8]
## Summary

Integrate all components into a complete, usable system with CLI, configuration, and documentation. This phase delivers a working end-to-end anomaly detection tool.

**Part of:** Meta Issue #1
**Dependencies:** All previous phases (1-6)

## Objectives

1. Create CLI entry point with argument parsing
2. Wire all components into a unified pipeline
3. Complete configuration with validation
4. Add comprehensive error handling and logging
5. Update documentation

## Files to Create/Update

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/main.py` | 150 | Create | CLI entry point |
| `src/pipeline.py` | 200 | Create | Full pipeline orchestration |
| `config/default.yaml` | 100 | Update | Complete configuration |
| `requirements.txt` | 20 | Update | All dependencies |
| `.gitignore` | 30 | Create | Git ignore patterns |
| `tests/test_integration.py` | 150 | Create | Integration tests |
| `README.md` | - | Update | Usage instructions |

### CLI Entry Point
- [ ] Add argparse with --config, --input, --output, --sample options
- [ ] Support batch mode (process file) and streaming mode
- [ ] Add verbose/quiet output control
- [ ] Handle keyboard interrupt gracefully

### Pipeline Orchestration
- [ ] Create `AnomalyPipeline` coordinating all components
- [ ] Implement `run_batch()` for file processing
- [ ] Implement `run_streaming()` for real-time
- [ ] Add progress reporting for batch mode
- [ ] Save results to output directory

### Configuration
- [ ] Complete default.yaml with all sections
- [ ] Add YAML schema validation
- [ ] Support config overrides via CLI
- [ ] Document all configuration options

### Error Handling & Logging
- [ ] Set up Python logging with file and console handlers
- [ ] Log at appropriate levels (DEBUG, INFO, WARNING, ERROR)
- [ ] Wrap main execution in try/except
- [ ] Produce helpful error messages

### Documentation
- [ ] Update README with installation and usage
- [ ] Add example commands
- [ ] Document configuration options
- [ ] Add troubleshooting section

## Code Snippet: CLI Entry Point

```python
# src/main.py
"""Command-line interface for anomaly detection system."""

import argparse
import logging
import sys
from pathlib import Path

import yaml

from .pipeline import AnomalyPipeline


def setup_logging(config: dict) -> None:
    """Configure logging from config."""
    log_config = config.get('logging', {})
    level = getattr(logging, log_config.get('level', 'INFO'))
    log_file = log_config.get('file')

    handlers = [logging.StreamHandler(sys.stdout)]
    if log_file:
        Path(log_file).parent.mkdir(parents=True, exist_ok=True)
        handlers.append(logging.FileHandler(log_file))

    logging.basicConfig(
        level=level,
        format='%(asctime)s [%(levelname)s] %(name)s: %(message)s',
        handlers=handlers
    )


def load_config(config_path: str) -> dict:
    """Load YAML configuration."""
    with open(config_path, 'r') as f:
        return yaml.safe_load(f)


def main():
    parser = argparse.ArgumentParser(
        description='High-Dimensional Anomaly Detection System'
    )
    parser.add_argument(
        '--config', '-c',
        default='config/default.yaml',
        help='Path to configuration file'
    )
    parser.add_argument(
        '--input', '-i',
        help='Input data file (CSV)'
    )
    parser.add_argument(
        '--output', '-o',
        default='results/',
        help='Output directory for results'
    )
    parser.add_argument(
        '--sample',
        type=float,
        help='Fraction of data to sample (0.0-1.0)'
    )
    parser.add_argument(
        '--mode',
        choices=['batch', 'streaming'],
        default='batch',
        help='Processing mode'
    )
    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='Enable verbose output'
    )

    args = parser.parse_args()

    # Load configuration
    config = load_config(args.config)

    # Override log level if verbose
    if args.verbose:
        config.setdefault('logging', {})['level'] = 'DEBUG'

    # Setup logging
    setup_logging(config)
    logger = logging.getLogger(__name__)

    try:
        logger.info("Starting anomaly detection pipeline")
        logger.info(f"Config: {args.config}")
        logger.info(f"Input: {args.input}")
        logger.info(f"Mode: {args.mode}")

        # Create pipeline
        pipeline = AnomalyPipeline(config)

        # Run
        if args.mode == 'batch':
            results = pipeline.run_batch(
                input_file=args.input,
                output_dir=args.output,
                sample_frac=args.sample
            )
        else:
            results = pipeline.run_streaming(
                input_file=args.input,
                output_dir=args.output
            )

        logger.info(f"Processing complete. Results saved to {args.output}")
        logger.info(f"Total anomalies detected: {results['anomaly_count']}")

    except KeyboardInterrupt:
        logger.info("Interrupted by user")
        sys.exit(1)
    except Exception as e:
        logger.error(f"Error: {e}", exc_info=True)
        sys.exit(1)


if __name__ == '__main__':
    main()
```

## Code Snippet: Pipeline Orchestration

```python
# src/pipeline.py
"""Full anomaly detection pipeline."""

import logging
from pathlib import Path
from typing import Dict, Optional
import json

import numpy as np
import pandas as pd

from .ingestion.data_loader import DataLoader
from .ingestion.preprocessor import Preprocessor
from .ingestion.stream_processor import StreamProcessor
from .reduction.autoencoder import AutoencoderReducer
from .detection.ensemble import AnomalyEnsemble
from .explanation.explainer import AnomalyExplainer
from .feedback.manager import FeedbackManager


class AnomalyPipeline:
    """Orchestrate full anomaly detection workflow."""

    def __init__(self, config: dict):
        self.config = config
        self.logger = logging.getLogger(__name__)

        # Initialize components
        self.loader = DataLoader(config)
        self.preprocessor = Preprocessor(config)
        self.reducer = AutoencoderReducer(config)
        self.ensemble = AnomalyEnsemble(config)
        self.explainer = None  # Created after fitting
        self.feedback = FeedbackManager(config)

        self.threshold = config.get('detection', {}).get('ensemble', {}).get('threshold', 0.7)

    def run_batch(
        self,
        input_file: str,
        output_dir: str,
        sample_frac: Optional[float] = None
    ) -> Dict:
        """
        Process file in batch mode.

        Returns:
            Dict with processing statistics
        """
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)

        # Load data
        self.logger.info(f"Loading data from {input_file}")
        df = self.loader.load_csv(input_file, sample_frac=sample_frac)
        self.logger.info(f"Loaded {len(df)} samples, {len(df.columns)} features")

        # Preprocess
        self.logger.info("Preprocessing data")
        X = self.preprocessor.fit_transform(df.values)

        # Reduce dimensions
        self.logger.info("Reducing dimensions")
        self.reducer.fit(X)
        X_reduced = self.reducer.transform(X)
        self.logger.info(f"Reduced to {X_reduced.shape[1]} dimensions")

        # Detect anomalies
        self.logger.info("Detecting anomalies")
        self.ensemble.fit(X_reduced)
        scores = self.ensemble.score_samples(X_reduced)

        # Create explainer
        self.explainer = AnomalyExplainer(self.config, self.ensemble, self.reducer)

        # Get anomalies
        is_anomaly = scores > self.threshold
        anomaly_indices = np.where(is_anomaly)[0]
        self.logger.info(f"Found {len(anomaly_indices)} anomalies ({len(anomaly_indices)/len(X)*100:.1f}%)")

        # Generate explanations for anomalies
        if len(anomaly_indices) > 0:
            self.logger.info("Generating explanations")
            X_anomalies = X[anomaly_indices]
            explanations = self.explainer.explain(
                X_anomalies,
                self.threshold,
                sample_indices=anomaly_indices.tolist()
            )

            # Save explanations
            explanations_data = [e.to_dict() for e in explanations]
            with open(output_path / 'explanations.json', 'w') as f:
                json.dump(explanations_data, f, indent=2)

        # Save all scores
        results_df = pd.DataFrame({
            'sample_index': range(len(scores)),
            'anomaly_score': scores,
            'is_anomaly': is_anomaly
        })
        results_df.to_csv(output_path / 'scores.csv', index=False)

        return {
            'total_samples': len(X),
            'anomaly_count': len(anomaly_indices),
            'anomaly_rate': len(anomaly_indices) / len(X),
            'threshold': self.threshold
        }

    def run_streaming(self, input_file: str, output_dir: str) -> Dict:
        """Process file in streaming mode."""
        # Initialize with first chunk, then process stream
        # Implementation similar to StreamProcessor usage
        self.logger.info("Streaming mode not fully implemented yet")
        return {'status': 'not_implemented'}
```

## Code Snippet: Complete Configuration

```yaml
# config/default.yaml

data:
  path: "data/"
  default_file: "input.csv"

preprocessing:
  normalize: true
  handle_missing: "median"

reduction:
  kernel_pca:
    n_components: 20
    kernel: "rbf"
    gamma: 0.1
  autoencoder:
    n_components: 10
    hidden_size: 64
    max_iter: 500
  incremental:
    n_components: 15
    batch_size: 1000

detection:
  isolation_forest:
    n_estimators: 100
    contamination: 0.1
    random_state: 42
  lof:
    n_neighbors: 20
    contamination: 0.1
  spc:
    method: "mahalanobis"
    regularization: 0.01
  ensemble:
    weights:
      isolation_forest: 1.0
      lof: 1.0
      spc: 0.8
    threshold: 0.7

streaming:
  window_size: 1000
  stride: 100
  max_buffer: 50000

explanation:
  top_k: 5
  method: "perturbation"

feedback:
  storage:
    db_path: "data/feedback.db"
  adaptation:
    learning_rate: 0.1
    min_samples: 10
  retrain:
    trigger_threshold: 100
    drift_threshold: 0.2

logging:
  level: "INFO"
  file: "logs/anomaly_detection.log"
```

## Verification

```bash
# Install dependencies
pip install -r requirements.txt

# Run with sample data
python -m src.main --config config/default.yaml --input data/sample.csv --output results/ --verbose

# Expected output:
# - results/scores.csv with all anomaly scores
# - results/explanations.json with anomaly explanations
# - logs/anomaly_detection.log with processing log

# Run tests
pytest tests/ -v

# Run integration test
pytest tests/test_integration.py -v
```

## Definition of Done

- [ ] CLI processes sample data without errors
- [ ] Configuration controls all parameters
- [ ] Errors produce helpful messages
- [ ] Logging captures key events
- [ ] Results saved to output directory
- [ ] README updated with usage instructions
- [ ] All tests pass

## Learning Notes

**Concepts to understand:**
- argparse for CLI argument parsing
- YAML configuration management
- Python logging module
- Exception handling patterns
- JSON serialization
